Download charts to a temporary file

amir opened the pull request:
Similar to https://github.com/kubernetes/helm/issues/3253 we're using
Helm by directly importing the Go package.

A concurrency issue arises when multiple requests come for the same
package. Currently DownloadTo truncates contents of destination file and
rewrites it.

This patch changes the behaviour so instead of directly writing into the
destination it'll write the content in a temp file and renames it once
download is finished.

This change fixes sporadic issues in
https://github.com/amir/skeg/blob/5c5c6a1c215cf5110843a28f1fa389aa69a3f452/main_test.go

amir committed:
Ask ioutil to write in a temp file

Similar to https://github.com/kubernetes/helm/issues/3253 we're using
Helm by directly importing the Go package.

A concurrency issue arises when multiple requests come for the same
package. Currently DownloadTo truncates contents of destination file and
rewrites it.

This patch changes the behaviour so instead of directly writing into the
destination it'll write the content in a temp file and renames it once
download is finished.

This change fixes sporadic issues in
https://github.com/amir/skeg/blob/5c5c6a1c215cf5110843a28f1fa389aa69a3f452/main_test.go

amir committed:
Fix error handling

technosophos reviewed (commented):
Given that `tempdestfile` is a hardcoded file, aren't we likely to run into this problem again, but with `.part` getting truncated? Maybe `ioutil.TempFile()` would save us from that?

amir committed:
Use ioutil.TempFile

amir commented:
Using `ioutil.TempFile()` now.

mattjmcnaughton commented on pkg/downloader/chart_downloader.go:
```diff
@@ -97,7 +97,16 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+	tmpfile, err := ioutil.TempFile(dest, "tmp")
+	if err != nil {
+		return "", nil, err
+	}
```
If there is a failure between the creation of the tmpfile and the `os.Rename` command, I believe that the `tmpfile` will never be removed (because it is the user's responsibility to clean up files created by `ioutil.TempFile`). Not sure if this is an actual problem, but if so, I think we could solve with a 

```
defer func() {
    if _, err := os.State(tmpfile.Name()); err == nil {
        os.Remove(tmpfileName())
    }
}

mattjmcnaughton commented on pkg/downloader/chart_downloader.go:
```diff
@@ -97,7 +97,16 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+		return "", nil, err
+	}
+	tmpfilename := tmpfile.Name()
+	if err := ioutil.WriteFile(tmpfilename, data.Bytes(), 0644); err != nil {
```
I believe that the `0644` permissions will only be applied if `ioutil.WriteFile` creates a file which does not already exist. Since we create `tmpfile` with `TempFile`, I believe that we need to use `os.Chmod` to change the permissions of the `TempFile`.

mattjmcnaughton reviewed (changes requested):
Just two small thoughts from me, but otherwise looks good :)

amir committed:
Remove tempfile if necessary and chmod destfile

amir commented:
@mattjmcnaughton thanks!

mattjmcnaughton commented on pkg/downloader/chart_downloader.go:
```diff
@@ -98,10 +98,20 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+
 	tmpfilename := tmpfile.Name()
+	// 0644 here is ineffective since TempFile creates files with 0600 permission. We'll chmod it later.
 	if err := ioutil.WriteFile(tmpfilename, data.Bytes(), 0644); err != nil {
```
Is it possible to just do `tmpfile.Write(data.Bytes())`? imho, `ioutil.WriteFile` is adding some confusion that isn't necessary around the permissioning.

mattjmcnaughton reviewed (approved):
LGTM modulo the comment on `ioutil.WriteFile` vs `tmpfile.Write`. Feel free to resolve however you think is best :)

amir committed:
Replace ioutil.WriteFile with File.Write

This would eliminate the need to comment the need to Chmod.
However, we need to Close the file manually now since ioutil.WriteFile
does it internally.

bacongobbler commented on pkg/downloader/chart_downloader.go:
```diff
@@ -97,7 +97,32 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+	defer func() {
+		if tmpfile != nil {
+			if _, err := os.Stat(tmpfile.Name()); err == nil {
+				os.Remove(tmpfile.Name())
```
on Windows it won't remove the file unless we release the handler to the file. Can we call `tmpfile.Close()` here? There's a chance we can return on a bad write, causing the handle to never be released.

bacongobbler commented on pkg/downloader/chart_downloader.go:
```diff
@@ -97,7 +97,32 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+	}
+
+	tmpfilename := tmpfile.Name()
+	if _, err := tmpfile.Write(data.Bytes()); err != nil {
```
generally speaking the more idiomatic Go approach is to call `io.Copy(tmpfile, data)` instead. Can we make that change here?

amir committed:
Close tmpfile before removing it

Also replace tmpfile.Write with io.Copy

amir commented:
@bacongobbler thanks for the review. I made the changes.

bacongobbler commented:
Sorry for not getting back sooner.

Can we use a mutex or a file lock instead of a tempfile write/relocate? See https://github.com/helm/helm/pull/6492.

karuppiah7890 commented:
This would also fix #6140 I guess?

bacongobbler commented:
seems plausible

bacongobbler commented:
I'm going to close this ticket due to inactivity, but please re-open if this still needs to be addressed. Thanks!
